### DISCUSSION TITLE
migrating/squirrel

### DISCUSSION THREAD

**Discussion by giscus** [velopack/velopack.docs#16] [General]
# migrating/squirrel

Migrate Squirrel.Windows or Clowd.Squirrel to Velopack.

https://docs.velopack.io/migrating/squirrel

**Comment by mgnslndh** [comment]
Can a Velopack installer upgrade an app installed using Squirrel?

**Reply by caesay** [reply]
Yes, if you follow the steps in this guide you should be able to migrate an installed app from Squirrel to Velopack. Reach out via GitHub issues or Discord if you have any issues.

**Comment by shunkica** [comment]
What is the Velopack alternative to Squirrels `UpdateManager.RestartApp()`?
I am not trying to update the app, just trigger a restart.

**Reply by caesay** [reply]
There is no basic "Restart app but don't try and apply pending updates" functionality in Velopack. If you need to restart for reasons unrelated to updating I'd suggest doing something like the following:
```cs
if (Process.Start(Process.GetCurrentProcess().MainModule!.FileName) != null) {
    Environment.Exit(0);
}
```

**Comment by YoelRT** [comment]
Hello, my app is build using <i>msbuild</i> (frequently using <i>DevOps</i> ) there is a place you can doc or explain how to install/run <i>vpk</i> if not yet installed in the pipeline? I see another nuget package aside of <i>Velopack</i>, its name is <i>Velopack.build</i>, what's that used for?

**Reply by caesay** [reply]
The procedure to install vpk is the same regardless of where you use Velopack or what programming language you use. To install vpk you need to have dotnet8 installed, and then run `dotnet tool update -g vpk`. If your CI runner does not have dotnet installed, you may need to install it first. `Velopack.Build` is not documented or supported yet but may make things easier for msbuild in the future.

**Comment by YoelRT** [comment]
Hello, I am migrating from <i>Squirrel</i>, so in the whole process of creating a pack/installer, our building process uses a <i>client.nuspec</i> file to specify which files copy from the build folder to the installer. Is there any option we got for that aside of the <i>vpk pack --exclude</i> flag?

**Reply by caesay** [reply]
It's not supported to use a `.nuspec` or give Velopack a `.nupkg`. Ideally you should configure msbuild to output your project structure in the correct way, and copy or delete files as needed before giving your program output to Velopack. Your program will be packaged exactly how you give it to Velopack's `--packDir` arg, including preserving any symlinks etc. If you can't achieve what you need with msbuild, you might want to create a `build.bat` which prepares your application for packaging and then calls `vpk pack` afterwards. You could adapt one of the build scripts [from the samples](https://github.com/velopack/velopack/blob/develop/samples/CSharpWpf/build.bat).

**Comment by JamCrackersBV** [comment]
Hi Caesay,

Thank you for this wonderfull project. Im migrating my squirrel updated apps to Velopack.
I have one question regarding the migration. if an app is installed already in %localappdata%\AppDirectory and it gets updated.. although as appID in the vpk pack command I've choosen something else, the app still would be running in the 'old direcory' (in this example [AppDirectory], is that correct. So only n a fresh install it would 'change' directory (and actually use my appID al directory name) ?
Thanks

**Reply by caesay** [reply]
You are correct

**Reply by JamCrackersBV** [reply]
thnks!

**Comment by YoelRT** [comment]
Hello there, I am almost finishing the *Squirrel* replacement, however we are facing an issue while installing the app, a dialog shows: *"Installation has completed, but the application install hook failed ..."* I don't get any issue in the event-viewer and installing the *Setup.exe -v -l c:\temp\velo.log* shows not errors but a warning.
<i>
[setup:30972] [12:37:12] [INFO] Starting process install hook
[setup:30972] [12:37:12] [INFO] Running --veloapp-install hook...
[setup:30972] [12:37:27] **[WARN] Hook exited with non-zero exit code: -532462766**
[setup:30972] [12:37:27] [INFO] Checking for running processes in: C:\Users\user1\AppData\Local\CompanyApp
[setup:30972] [12:37:39] [INFO] Writing uninstall registry key...
[setup:30972] [12:37:39] [INFO] Starting app...
[setup:30972] [12:37:39] [DEBUG] (1) velopack_bins::shared::util_windows: Setting environment variable: VELOPACK_FIRSTRUN=true
[setup:30972] [12:37:39] [INFO] About to launch: '"C:\\Users\\user1\\AppData\\Local\\CompanyApp\\current\\App.exe"' in dir '"C:\\Users\\user1\\AppData\\Local\\CompanyApp\\current"'
[setup:30972] [12:37:39] [INFO] Args: CommandArgs { inner: [] }
[setup:30972] [12:37:39] [INFO] Installation completed successfully!
[setup:30972] [12:37:39] [INFO] Removing rollback directory...
</i>
Any idea how to find the issue?

**Reply by caesay** [reply]
> Hook exited with non-zero exit code: -532462766

Your own application crashed during install and/or did not return a 0 exit code. I suggest you add more logging or exception handling to your product when it starts up. Also make sure you are calling `VelopackApp.Run()` before anything else in your app startup.

Comments on our documentation are not the best place for extended conversations or support - the comments here are designed for asking questions or discussing the documentation itself. If you need support, please open an issue on GitHub or join our Discord.

**Comment by C1rdec** [comment]
After the migration the new process (With Velopack) get stuck.... But if I kill it and restart the app everything goes back to normal.

Is this a known issue or is there a workaround?

**Reply by C1rdec** [reply]
```
2024-11-30 14:20:46.7963 INFO Initializing WindowsVelopackLocator
2024-11-30 14:20:46.8049 INFO Update.exe found in parent directory
2024-11-30 14:20:46.8049 INFO Located valid manifest file at: AppData\Local\PoeLurker\current\sq.version
2024-11-30 14:20:46.8049 INFO Starting Velopack App (Run).
2024-11-30 14:20:46.8049 INFO Setting current process explicit AppUserModelID to 'velopack.PoeLurker'
```

**Reply by caesay** [reply]
Not a known issue. Please join our Discord or open a GitHub issue with migration logs for assistance.

**Comment by pankaj-nikam** [comment]
How can I use Velvopack with .NET framework 4.6.2 apps?

**Reply by caesay** [reply]
Please follow our [Quick Start Guide for C#](https://docs.velopack.io/getting-started/csharp), replacing `dotnet publish` with `msbuild` or whatever you currently use to compile your app. Also, review [our WPF sample](https://github.com/velopack/velopack/tree/develop/samples/CSharpWpf) which currently targets `net48`.

**Reply by pankaj-nikam** [reply]
Thank you so much @caesay :)
